hw/ip/otbn/util/shared/control_flow.py

```diff
@@ -0,0 +1,380 @@
+        for pc in edges.keys():
+            assert pc in sections
+
+        self.start = start
+        self.sections = sections
+        self.edges = edges
+
+    def get_section(self, pc) -> CodeSection:
+        '''Returns the CodeSection for the given PC.
+
+        Raises a ValueError if the PC is not in the graph.
+        '''
+        if pc not in self.sections:
+            raise ValueError(
+                'PC {:#x} not found in control graph (PCs in graph are: {})'.
+                format(pc,
+                       ['{:#x}'.format(pc) for pc in self.sections.keys()]))
+        return self.sections[pc]
+
+    def get_edges(self, pc) -> List[ControlLoc]:
+        '''Returns the edges for the section starting at pc.
+
+        Raises a ValueError if the PC is not in the graph.
+        '''
+        if pc not in self.edges:
+            raise ValueError(
+                'PC {:#x} not found in control graph (PCs in graph are: {})'.
+                format(pc, ['{:#x}'.format(pc) for pc in self.edges.keys()]))
+        return self.edges[pc]
+
+    def get_cycles(self, pc, path=[]) -> Dict[int, List[int]]:
+        '''Returns any control-flow cycles in this graph.'''
+        if pc not in self.sections:
+            raise ValueError(
+                'Invalid entry PC when getting cycles: {:#x}. Valid '
+                'control PCs are: {}'.format(
+                    pc, ['{:#x}'.format(pc) for pc in self.sections.keys()]))
+        cycles: Dict[int, List[int]] = {}
+        sec = self.get_section(pc)
+        for loc in self.get_edges(pc):
+            if isinstance(loc, LoopStart):
+                # If this is the last loop instruction, don't loop back
+                if sec.end == loc.loop_end_pc:
+                    continue
+                pc = loc.loop_start_pc
+            elif loc.is_special():
+                # Skip Ret, Ecall, ImemEnd
+                continue
+            else:
+                pc = loc.pc
+            if pc in path:
+                # Found a cycle; report it
+                if pc not in cycles:
+                    cycles[pc] = []
+                cycles[pc].append(sec.end)
+            else:
+                cycles.update(
+                    self.get_cycles(
+                        pc, path + list(range(sec.start, sec.end + 4, 4))))
+        return cycles
+
+    def _pretty_lines(self,
+                      program,
+                      indent,
+                      entry_pc,
+                      concise,
+                      already_printed,
+                      call_stack=[]):
+        '''Returns lines for pretty-printing.'''
+        if entry_pc not in self.sections:
+            raise ValueError(
+                'Invalid entry PC when pretty-printing control graph: '
+                '{:#x}. Valid control PCs are: {}'.format(
+                    entry_pc,
+                    ['{:#x}'.format(pc) for pc in self.sections.keys()]))
+        out = []
+        prefix = ' ' * indent
```

Comment:
i kinda wonder if the type of this function aught to be `[(int, str)]`, where the integer is the indent level...

Reply:
Changed, let me know if this is better; if so I'll change the one in #9627 also.